### PR TITLE
fix(api): use `text_locked()` to check for textlock

### PR DESCRIPTION
**Pull request by seandewar**
tl;dr:
- fixes various ways of escaping or screwing up the cmdwin via the api (and termopen)
- adds textlock checks to a few api functions missing them (and termopen); improves error in #21055 as a result
- fixes regression in #22634 allowing textlock-checking api functions to be usable from `<expr>` maps

idk if this is worth the new api attribute :shrug:, but changes `FUNC_API_CHECK_TEXTLOCK` to also check cmdwin by default (via `text_locked()`), and adds `FUNC_API_TEXTLOCK_ALLOW_CMDWIN` as an opt-in for the old behaviour which does not check cmdwin. Also, `FUNC_API_CHECK_TEXTLOCK` is now renamed to `FUNC_API_TEXTLOCK`.

The idea of opting-in to the old behaviour is to reduce the chance of reintroducing these bugs for new `curwin`/`curbuf`-switching APIs added in the future (this default may be overly restrictive for some APIs though, like those which strictly just change the buffer text).

### CODE REVIEW

File: src/nvim/func_attr.h [hunk 1]
```diff
@@ -228,6 +228,8 @@
 # define FUNC_API_LUA_ONLY
 /// API function checked textlock.
 # define FUNC_API_CHECK_TEXTLOCK
+/// API function checked textlock, but still callable in the cmdwin.
+# define FUNC_API_CHECK_TEXTLOCK_ALLOW_CMDWIN
```
Comment (justinmk) [inline src/nvim/func_attr.h:232]:
```suggestion
# define FUNC_API_TEXTLOCK
/// API function fails during textlock, but allows cmdwin.
# define FUNC_API_TEXTLOCK_ALLOW_CMDWIN
```

Comment (justinmk) [inline src/nvim/func_attr.h:232]:
```suggestion
/// API function fails during textlock.
# define FUNC_API_TEXTLOCK
/// API function fails during textlock, but allows cmdwin.
# define FUNC_API_TEXTLOCK_ALLOW_CMDWIN
```

File: test/functional/terminal/buffer_spec.lua [hunk 2]
```diff
@@ -498,3 +498,17 @@ if is_os('win') then
     end)
   end)
 end
+
+describe('termopen()', function()
```
Comment (justinmk) [inline test/functional/terminal/buffer_spec.lua:502]: 👍   looks like we don't have an existing describe() for termopen()

File: src/nvim/api/vim.c [hunk 3]
```diff
@@ -951,12 +951,18 @@ Buffer nvim_create_buf(Boolean listed, Boolean scratch, Error *err)
...
   FUNC_API_SINCE(7)
+  FUNC_API_CHECK_TEXTLOCK_ALLOW_CMDWIN
 {
   buf_T *buf = find_buffer_by_handle(buffer, err);
   if (!buf) {
     return 0;
   }
 
+  if (cmdwin_type != 0 && buf == curbuf) {
+    api_set_error(err, kErrorTypeException, "%s", _(e_cmdwin));
```
Comment (justinmk) [inline src/nvim/api/vim.c:962]: Does this disallow `nvim_open_term` in cmdwin? If so, why is it marked as FUNC_API_CHECK_TEXTLOCK_ALLOW_CMDWIN ?

Comment (seandewar) [reply src/nvim/api/vim.c:962]: It's allowed in the cmdwin, as long as you don't try to open a terminal in the cmdwin's buffer.

File: src/nvim/generators/c_grammar.lua [hunk 4]
```diff
@@ -47,7 +47,8 @@ local c_proto = Ct(
   (fill * Cg((P('FUNC_API_NOEXPORT') * Cc(true)), 'noexport') ^ -1) *
   (fill * Cg((P('FUNC_API_REMOTE_ONLY') * Cc(true)), 'remote_only') ^ -1) *
   (fill * Cg((P('FUNC_API_LUA_ONLY') * Cc(true)), 'lua_only') ^ -1) *
-  (fill * Cg((P('FUNC_API_CHECK_TEXTLOCK') * Cc(true)), 'check_textlock') ^ -1) *
+  (fill * (Cg(P('FUNC_API_CHECK_TEXTLOCK_ALLOW_CMDWIN') * Cc(true), 'check_textlock_allow_cmdwin') +
+           Cg(P('FUNC_API_CHECK_TEXTLOCK') * Cc(true), 'check_textlock')) ^ -1) *
```
Comment (justinmk) [inline src/nvim/generators/c_grammar.lua:51]: Would it make sense to instead have:

```
FUNC_API_TEXTLOCK
FUNC_API_NO_CMDWIN
```

So `FUNC_API_TEXTLOCK` has the "old" behavior of allow cmdwin. But functions that opt-in to allowing cmdwin are also marked `FUNC_API_NO_CMDWIN`.

Comment (seandewar) [reply src/nvim/generators/c_grammar.lua:51]: Do you mean `FUNC_API_NO_CMDWIN` should disallow cmdwin?

If so, I'm happy to do it that way, but people'll need to remember to add the other attribute when creating new APIs that could potentially break the cmdwin.

Comment (justinmk) [reply src/nvim/generators/c_grammar.lua]: > I'm happy to do it that way, but people'll need to remember to add the attribute when creating new APIs that could potentially break the cmdwin (otherwise it's a bug if they forget)

Either way is fine. Is this better:

- `FUNC_API_TEXTLOCK` implies "no cmdwin"
- `FUNC_API_CMDWIN` opts-in to allowing cmdline

**Review by justinmk — APPROVED** [review]
LGTM, let's give others a few days before merging

File: src/nvim/generators/c_grammar.lua (hunk 4 above)
Comment (justinmk) [reply src/nvim/generators/c_grammar.lua]: Current way looks fine. seem a bit weird to have textlock and cmdwin coupled, but the approach in your PR doesn't look like tech debt, it's easy to revisit it later.

Comment (seandewar) [reply src/nvim/generators/c_grammar.lua]: Yeah, admittedly it does feel a bit weird. (And it hurts my head! 😅) 

The reason they're related at all is if something cannot work in the cmdwin, it strongly implies it changes curwin/curbuf (as that is the main restrictions in the cmdwin), which also is not allowed during textlock.
The opposite is not quite as true though; if something cannot work during textlock, it might be that it only edits the buffer text, which is still allowed in the cmdwin.

File: test/functional/vimscript/api_functions_spec.lua [hunk 5]
```diff
@@ -48,10 +49,46 @@ describe('eval-API', function()
...
        pcall_err(command, "normal! yy"))
+
+    command("autocmd TextYankPost <buffer> ++once call nvim_open_term(0, {})")
+    matches('Vim%(call%):E5555: API call: E565: Not allowed to change text or change window$',
+       pcall_err(command, "normal! yy"))
+
+    -- Functions checking textlock should also not be usable from <expr> mappings.
+    command("nnoremap <expr> <f2> nvim_win_close(0, 1)")
+    eq('Vim(normal):E5555: API call: E565: Not allowed to change text or change window',
+       pcall_err(command, [[execute "normal \<f2>"]]))
```
Comment (zeertzjq) [inline test/functional/vimscript/api_functions_spec.lua:64]:
```suggestion
    command("inoremap <expr> <f2> nvim_win_close(0, 1)")
    eq('Vim(normal):E5555: API call: E565: Not allowed to change text or change window',
       pcall_err(command, [[execute "normal i\<f2>"]]))
```

File: test/functional/vimscript/api_functions_spec.lua [hunk 6]
```diff
@@ -48,10 +49,46 @@ describe('eval-API', function()
...
+
+    -- Text-changing functions gave a "Failed to save undo information" error when called from an
+    -- <expr> mapping outside do_cmdline() (msg_list == NULL), so use feed() to test this.
+    command("nnoremap <expr> <f2> nvim_buf_set_text(0, 0, 0, 0, 0, ['hi'])")
+    meths.set_vvar('errmsg', '')
+    feed("<f2>")
+    eq('E5555: API call: E565: Not allowed to change text or change window',
+       meths.get_vvar('errmsg'))
```
Comment (zeertzjq) [inline test/functional/vimscript/api_functions_spec.lua:72]:
```suggestion
    command("inoremap <expr> <f2> nvim_buf_set_text(0, 0, 0, 0, 0, ['hi'])")
    meths.set_vvar('errmsg', '')
    feed("i<f2><esc>")
    eq('E5555: API call: E565: Not allowed to change text or change window',
       meths.get_vvar('errmsg'))
```

**Review by zeertzjq — COMMENTED** [review]
I'd suggest using `:inoremap` to prevent the return value from doing strange things in Normal mode.

File: test/functional/vimscript/api_functions_spec.lua (hunk 5 above)
Comment (seandewar) [reply test/functional/vimscript/api_functions_spec.lua:64]: Good ideas, thanks! I'll apply them in a jiffy. :+1:

**Review by zeertzjq — APPROVED** [review]
